### PR TITLE
Experimental http2 adapter

### PR DESCRIPTION
**Pull request by yasuf**
###  Experimental HTTP2 Adapter
* This is exploratory, hopefully we can build a real adapter from this, I'll keep adding bits and pieces but I was able to test an http2 request using axios with this, it's very barebones and needs clean up  and plenty of more functionality before being ready for an alpha
* Related to https://github.com/axios/axios/issues/1175
* To test HTTP2 requests you'll need to:
  * Run `openssl req -x509 -newkey rsa:2048 -nodes -sha256 -subj '/CN=localhost' \
  -keyout localhost-privkey.pem -out localhost-cert.pem` which will generate the two files needed to run `http2_test.js`
  * After running that command you should see `localhost-privkey.pem` and `localhost-cert.pem` in your folder
  * In a terminal tab run `server.js` and in another tab run `http2_test.js` which will immediately send a request to the server

Feel free to comment on this PR, feedback is welcome

### CODE REVIEW

File: http2_test.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+var axios = require('./index.js');
+var fs = require('fs');
+
+axios.defaults.experimental_http2 = true;
```
Comment (yasuf) [inline http2_test.js]: this line doesn't do anything right now, it's sort of the idea, to have some kind of config boolean that can enable http2

File: lib/adapters/http2.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    var path = buildURL(parsed.path, config.params, config.paramsSerializer).replace(/^\?/, '');
+    var client = http2.connect(parsed.href, {
+      ca: [fs.readFileSync(config.certificatePath).toString()]
+    });
+    var protocol = parsed.protocol || 'http:';
+    var req = client.request({
+      ':path': parsed.path,
+    });
+
+    client.on('error', (err) => {
```
Comment (Alanscut) [inline lib/adapters/http2.js]: currently does not support the arrow function,maybe we can introduce babel

File: lib/defaults.js [hunk 3]
```diff
@@ -13,19 +15,31 @@ function setContentTypeIfUnset(headers, value) {
...
+ */
 function getDefaultAdapter() {
   var adapter;
   if (typeof XMLHttpRequest !== 'undefined') {
-    // For browsers use XHR adapter
     adapter = require('./adapters/xhr');
   } else if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
-    // For node use HTTP adapter
-    adapter = require('./adapters/http');
+    adapter = getNodeAdapter();
```
Comment (Alanscut) [inline lib/defaults.js]: although the function will improve,eslint suggests that we should use after define

File: lib/defaults.js [hunk 4]
```diff
@@ -13,19 +15,31 @@ function setContentTypeIfUnset(headers, value) {
...
   }
   return adapter;
 }
 
-var defaults = {
+function getNodeAdapter() {
+  // TODO: Add proper conditional based on defaults
+  if (true) {
```
Comment (Alanscut) [inline lib/defaults.js]: maybe what you mean is this:
```
if(true){
return require('./adapters/http2')';
}
return require('./adapters/http');
```

**Comment by yasuf** [comment]
now it's a little better where people can start trying it out if they need to

**Comment by yasuf** [comment]
thanks @Alanscut for taking a look1

**Comment by Alanscut** [comment]
@yasuf Anytime! 
I think the implementation of `http2` should be the same as `http`, so that the usage can be consistent.
When axios sets `maxRedirects`, the `httpAdapter` will initiate a request via `follow-redirects` to implement the redirect.
However, [follow-redirects](https://github.com/follow-redirects/follow-redirects) does not support http2. I have submit an issue in the project, but not sure if the lib will support http2 in the future.

**Comment by Alanscut** [comment]
By the way, the CI failed is caused by eslint error. I use VS Code to develop code, and installed the eslint plugin to avoid eslint errors. Recommend you use VS Code + eslint plugin :)

**Comment by bkulyk** [comment]
@yasuf Why not publish this as a separate module. Doesn't seem to be any need to have this included in the core package.

**Comment by yasuf** [comment]
@bkulyk it certainly can be, there was a lot of interest in the community around supporting http2 natively in axios in https://github.com/axios/axios/issues/1175 but it's a good idea to keep it separate, don't know what's the preference though since I haven't used http2 yet

**Comment by bkulyk** [comment]
@yasuf yea there was interest, but it's not helping anybody if it doesn't get merged. I gave it a try and seems to be working, but would get more use if it were published somewhere.

**Comment by bkulyk** [comment]
Says it's out of date with the base and some checks have failed. Can you update so it has a chance to be merged

**Comment by silverbackdan** [comment]
Hi all - this looks really exciting as others have been discussing in the issue, but the CI checks never seemed to run. It'd be good to push the run through if possible and see if there's a way to get this merged and start experimenting?

**Comment by jasonsaayman** [comment]
@silverbackdan please could you run this locally and have a check that it does as it should.

**Comment by jasonsaayman** [comment]
@silverbackdan Seems that it fails testing immediately, could you have a look?

**Comment by silverbackdan** [comment]
Sure, it looks like easy lint issues - as soon as I have a moment I'll give it a go. Shall I just fork from @yasuf to continue with a new PR or if I contribute, how would be best for you?

**Comment by jasonsaayman** [comment]
Hi,

This seems to have lost steam as a pull request, but it was also experimental and seems very alpha at this stage. For that reason and since http2 seems to be far off I believe it would be better to close this for now and look at a project in the future where we can wholeheartedly commit to a rewrite with http2 included.

Please open a new issue or pull request if this issue comes up again.

Thanks

**Comment by zavoyevatyel** [comment]
When will the update be finished, presumably?

**Pull request closed without merging**